multicluster/controllers/multicluster/core/remote_cluster.go

```diff
@@ -0,0 +1,353 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *remoteCluster) updateRemoteClusterStatus(connected bool, err error) {
+	if r.connected == connected {
+		return
+	}
+
+	klog.Infof("Updating remote cluster status", "cluster", r.GetClusterID(), "connected", connected)
+
+	// TODO: Tolerate transient failures so we dont oscillate between connected and disconnected.
+	r.connected = connected
+}
+
+/**
+ * ---------------------------
+ * commonArea Implementation
+ * ---------------------------
+ */
+
+func (r *remoteCluster) GetClusterID() common.ClusterID {
+	return r.ClusterID
+}
+
+func (r *remoteCluster) GetNamespace() string {
+	return r.Namespace
+}
+
+/**
+ * ---------------------------
+ * RemoteCluster Implementation
+ * ---------------------------
+ */
+
+/**
+ * Once connected to the remote cluster, the start method run a timer
+ * on a go routine to periodically write MemberAnnounce CRD into the remote
+ * cluster's common area and also maintain its connectivity status to the
+ * remote cluster
+ */
+func (r *remoteCluster) Start() (context.CancelFunc, error) {
+	stopCtx, stopFunc := context.WithCancel(context.Background())
+
+	// Start a Timer for every 5 seconds
+	// TODO: make the internal longer? the webhook API is called every 5s to remote cluster now
+	ticker := time.NewTicker(5 * time.Second)
+
+	go func() {
+		klog.Infof("Starting MemberAnnounce", "cluster", r.GetClusterID())
+		for {
+			select {
+			case <-stopCtx.Done():
+				klog.Infof("Stopping MemberAnnounce", "cluster", r.GetClusterID())
+				return
+			case <-ticker.C:
+				// do member announce
+				if err := r.SendMemberAnnounce(); err != nil {
+					// will be tried again
+					klog.Error(err, "error writing member announce", "cluster", r.GetClusterID())
+					r.updateRemoteClusterStatus(false, err)
+				} else {
+					r.updateRemoteClusterStatus(true, nil)
+				}
+			}
+		}
+	}()
+
+	r.stopFunc = stopFunc
+	return stopFunc, nil
+}
+
+func (r *remoteCluster) Stop() error {
+	if r.stopFunc == nil {
+		return nil
+	}
+
+	r.stopFunc()
+	return nil
+}
+
+func (r *remoteCluster) IsConnected() bool {
+	return r.connected
+}
+
+func (r *remoteCluster) StartMonitoring() error {
+	klog.V(2).Info("Starting monitoring for resourceimports from", "remoteCluster", r.ClusterID)
+
+	resImportReconciler := NewResourceImportReconciler(
+		r.ClusterManager.GetClient(),
+		r.ClusterManager.GetScheme(),
+		r.localClusterManager,
+		r,
+	)
+	if err := resImportReconciler.SetupWithManager(r.ClusterManager); err != nil {
+		klog.V(2).ErrorS(err, "unable to create controller", "controller", "ResourceImport")
+		return fmt.Errorf("unable to create ResourceImport webhook, err: %v", err)
+	}
+	if err := (&multiclusterv1alpha1.ResourceImport{}).SetupWebhookWithManager(r.ClusterManager); err != nil {
+		klog.V(2).ErrorS(err, "unable to create webhook", "webhook", "ResourceImport")
+		return fmt.Errorf("unable to create ResourceImport webhook, err: %v", err)
+	}
+
+	go func() {
```

Comment:
Do we need a goroutine?

Reply:
This is the one that is writing MemberClusterAnnounce periodically. I can invoke it from RemoteClusterManager and avoid the goroutine, but then I need to expose a method like DoMemberClusterAnnounce(). Do you prefer that over the goroutine?

Reply:
What does r.ClusterManager.Start() do? I think MemberClusterAnnounce is by RemoteCluster.Start(), which is called elsewhile?

Reply:
oh sorry I dint look at the context properly. So this goroutine is not for periodically writing MemberClusterAnnounce, my bad.
This one does r.ClusterManager.Start() like you said and that is actually blocking until we stop it. It is the one that runs the controller-runtime Manager in which the ImportResource controller runs.

Reply:
So since it is blocking we do need a goroutine, right?

Reply:
When Start() returns? Could we call it twice if the leader is disconnected and reconnected? And what will happen then? I feel the complexity is we have another mechanism to check a leader is connected or not.

Probably we need some comments here.

Reply:
I did add a method now called StopWatching which will close the context so this blocking method can return. I also added a safety check so StartWatching is not called twice and added some comments, kindly let me know if its ok.

Reply:
Ok. I did not notice StopWatching will stop this too. Good to have some comments.